### PR TITLE
Object detection API calls in python

### PR DESCRIPTION
**Pull request by maciek-slon**
Related platform PR: https://github.com/rapp-project/rapp-platform/pull/346

### CODE REVIEW

File: python/RappCloud/CloudMsgs/ObjectDetectionClearModels.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        def __init__(self, **kwargs):
+            """!
+            Constructor
+
+            @param **kwargs - Keyword arguments. Apply values to the request attributes.
+                - @ref user
+            """
+
+            ## Username 
+            self.user = 'rapp'
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionClearModels.py]: To remove. User information is retrieved on the RAPP Platform using the user's credentials (token).
Resource [here](https://github.com/rapp-project/rapp-platform/wiki/How-to-create-a-HOP-service-for-a-ROS-service%3F#web-service-implementation)

File: python/RappCloud/CloudMsgs/ObjectDetectionClearModels.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            # Apply passed keyword arguments to the Request object.
+            super(ObjectDetectionClearModels.Request, self).__init__(**kwargs)
+
+
+        def make_payload(self):
+            """ Create and return the Payload of the Request. """
+            return Payload(user=self.user)
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionClearModels.py]: Remove `user` param from payload

File: python/RappCloud/CloudMsgs/ObjectDetectionLearnObject.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                - @ref name
+                - @ref user
+            """
+
+            ## File path to the image to load. This is the sample object image.
+            self.fname = ''
+            ## Model name
+            self.name = ''
+            ## Username 
+            self.user = 'rapp'
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionLearnObject.py]: To remove. User information is retrieved on the RAPP Platform using the user's credentials (token).
Resource [here](https://github.com/rapp-project/rapp-platform/wiki/How-to-create-a-HOP-service-for-a-ROS-service%3F#web-service-implementation)

File: python/RappCloud/CloudMsgs/ObjectDetectionLearnObject.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            # Apply passed keyword arguments to the Request object.
+            super(ObjectDetectionLearnObject.Request, self).__init__(**kwargs)
+
+
+        def make_payload(self):
+            """ Create and return the Payload of the Request. """
+            return Payload(name=self.name,user=self.user)
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionLearnObject.py]: Remove `user` param from payload

File: python/RappCloud/CloudMsgs/ObjectDetectionLoadModels.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+            @param **kwargs - Keyword arguments. Apply values to the request attributes.
+                - @ref imageFilepath
+                - @ref fast
+            """
+
+            ## Model names to be loaded
+            self.names = []
+            ## Username 
+            self.user = 'rapp'
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionLoadModels.py]: To remove. User information is retrieved on the RAPP Platform using the user's credentials (token).
Resource [here](https://github.com/rapp-project/rapp-platform/wiki/How-to-create-a-HOP-service-for-a-ROS-service%3F#web-service-implementation)

File: python/RappCloud/CloudMsgs/ObjectDetectionLoadModels.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            # Apply passed keyword arguments to the Request object.
+            super(ObjectDetectionLoadModels.Request, self).__init__(**kwargs)
+
+
+        def make_payload(self):
+            """ Create and return the Payload of the Request. """
+            return Payload(names=self.names,user=self.user)
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionLoadModels.py]: Remove `user` param from payload

File: python/RappCloud/RappPlatformApi.py [hunk 7]
```diff
@@ -840,7 +840,7 @@ def geolocation(self, ipaddr, engine='ip-api'):
         }
         
     def objectDetectionFindObjects(self, fname, limit = 1, user = 'rapp'):
```
Comment (klpanagi) [inline python/RappCloud/RappPlatformApi.py]: Remove `user` argument

File: python/RappCloud/RappPlatformApi.py [hunk 8]
```diff
@@ -868,3 +869,78 @@ def objectDetectionFindObjects(self, fname, limit = 1, user = 'rapp'):
             'result': response.result,
             'error': response.error
         }
+        
+    def objectDetectionClearModels(self, user = 'rapp'):
```
Comment (klpanagi) [inline python/RappCloud/RappPlatformApi.py]: Remove `user` argument

File: python/RappCloud/RappPlatformApi.py [hunk 9]
```diff
@@ -868,3 +869,78 @@ def objectDetectionFindObjects(self, fname, limit = 1, user = 'rapp'):
...
+            msg.req.user = user
+            response = self.svc_caller.call(msg)
+        except Exception as e:
+            response = ObjectDetectionClearModels.Response(error=str(e))
+        return {
+            'result': response.result,
+            'error': response.error
+        }
+        
+    def objectDetectionLoadModels(self, names = [], user = 'rapp'):
```
Comment (klpanagi) [inline python/RappCloud/RappPlatformApi.py]: Remove `user` argument

File: python/RappCloud/RappPlatformApi.py [hunk 10]
```diff
@@ -868,3 +869,78 @@ def objectDetectionFindObjects(self, fname, limit = 1, user = 'rapp'):
...
+            msg.req.user = user
+            response = self.svc_caller.call(msg)
+        except Exception as e:
+            response = ObjectDetectionLoadModels.Response(error=str(e))
+        return {
+            'result': response.result,
+            'error': response.error
+        }
+
+    def objectDetectionLearnObject(self, fname, name = '', user = 'rapp'):
```
Comment (klpanagi) [inline python/RappCloud/RappPlatformApi.py]: Remove `user` argument

File: python/RappCloud/CloudMsgs/ObjectDetectionFindObjects.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                - @ref fast
+            """
+
+            ## File path to the image to load. This is the image to perform
+            # object detection on.
+            self.fname = ''
+            ## Limit results to this number of objects
+            self.limit = 1
+            ## Username 
+            self.user = 'rapp'
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionFindObjects.py]: To remove. User information is retrieved on the RAPP Platform using the user's credentials (token).
Resource [here](https://github.com/rapp-project/rapp-platform/wiki/How-to-create-a-HOP-service-for-a-ROS-service%3F#web-service-implementation)

File: python/RappCloud/CloudMsgs/ObjectDetectionFindObjects.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+            # Apply passed keyword arguments to the Request object.
+            super(ObjectDetectionFindObjects.Request, self).__init__(**kwargs)
+
+
+        def make_payload(self):
+            """ Create and return the Payload of the Request. """
+            return Payload(limit=self.limit,user=self.user)
```
Comment (klpanagi) [inline python/RappCloud/CloudMsgs/ObjectDetectionFindObjects.py]: Remove `user` param from payload

File: python/RappCloud/RappPlatformApi.py [hunk 13]
```diff
@@ -837,3 +838,33 @@ def geolocation(self, ipaddr, engine='ip-api'):
             'zip': response.zip,
             'error': response.error
         }
+        
+    def objectDetectionFindObjects(self, fname, limit = 1, user = 'rapp'):
```
Comment (klpanagi) [inline python/RappCloud/RappPlatformApi.py]: Remove `user` argument

**Comment by klpanagi** [comment]
Hi @maciek-slon 

A few modifications are required in order to merge this PR:

- Remove `user` property from all `CloudMsg.Response` classes
- Remove `user`  argument from all RappPlatformAPI member methods

You can have a look [here](https://github.com/rapp-project/rapp-platform/wiki/How-to-create-a-HOP-service-for-a-ROS-service%3F#web-service-implementation) on how user authentication is performed using credentials.

**Comment by maciek-slon** [comment]
I need a way to have persistent storage for each user to store its models. At the moment `user` is used as a folder name under `rapp_platform_files` directory. Can I pass the token directly to ROS services (and use it as folder name), or is it security issue or sth?

**Comment by klpanagi** [comment]
O each call to the RAPP Platform services, a  token is required for authentication purposes. If you use the python API, the token is automatically loaded from `~/.config/rapp_platform/tokens/app` as described [here](https://github.com/rapp-project/rapp-api/tree/master/python#authenticationtokens)

On the Platform-side, on a request arrival, the request is first authenticated using that token. If authentication succeeds, the web service implementation callback is called and the `username` can be retrieved from the `req` object passed to the callback (`req.username`).

So in general, while using the python platform API, there is no need to send user information as those are automatically retrieved on the Platform and passed to the web service implementation (callback).

Also, there exist two web service implementation templates [here](https://github.com/rapp-project/rapp-platform/tree/master/rapp_web_services/services/templates)

I have placed a comment on the related rapp-platform PR [here](https://github.com/rapp-project/rapp-platform/pull/346/commits/cbf1ad3e0ba4e6cb91ff35944eb01d7550612404#r72674584)

Let me know if you need any further information.

**Comment by maciek-slon** [comment]
Username removed from service calls.

**Comment by klpanagi** [comment]
@maciek-slon  sorry for my late response. Can you to do this pull-request on branch **python** please; because this is related to the python implementation of the rapp-platform-api

**Comment by maciek-slon** [comment]
New PR: #51

**Pull request closed without merging**